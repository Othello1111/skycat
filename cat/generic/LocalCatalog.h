// -*-c++-*-
#ifndef _LocalCatalog_h_
#define _LocalCatalog_h_

/*
 * E.S.O. - VLT project/ESO Archive
 * $Id: LocalCatalog.h,v 1.1.1.1 2002/04/04 20:11:45 brighton Exp $
 *
 * LocalCatalog.h - class definitions for accessing local
 *                  catalogs stored as starbase format tab tables.
 *
 * See the man page for a complete description.
 *
 * who             when       what
 * --------------  --------   ----------------------------------------
 * Allan Brighton  11 Jun 96  Created
 * Peter W. Draper 21 Sep 98  Modified private data members to be
 *                            protected. Need these for derived
 *                            classes. 
 */

#include "AstroCatalog.h"

/*
 * Class LocalCatalog
 *
 * This class is used to search a local catalog stored as a file
 * in tab table format. The format of the table is something like:
 *
 * TableName
 *
 * VAR=Value
 *
 * A	B	C
 * -	-	-
 * 0	1	3
 * 3	2	4
 * ...
 *
 * where the table name and variable assignments are optional. 
 */
class LocalCatalog : public AstroCatalog {

protected:  // PWD: change here
    char* filename_;		// file name for local catalog
    time_t timestamp_;		// last update time of file, for caching

public:
    // constructor - create local catalog class instance
    // note: public interface uses AstroCatalog::open() with the name of the
    // file containing the local catalog.
    // The argument represents the entry in the catalog config file for this catalog
    // (made automatially, if not already present).
    LocalCatalog(CatalogInfoEntry*);

    // destructor
    ~LocalCatalog();

    // Run a query on the catalog and return the number of objects found.
    // (redefined here to work with local catalogs)
    virtual int query(const AstroQuery& q, const char* filename, QueryResult& result);

    // check the validity of a tab table file
    static int check_table(const char* file);

    // query server for catalog column names and put result in info_
    virtual int getInfo();
    virtual int checkInfo();
};

#endif /* _LocalCatalog_h_ */
