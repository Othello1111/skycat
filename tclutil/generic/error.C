/*
 * E.S.O. - VLT project 
 * "@(#) $Id: error.C,v 1.2 1998/02/18 11:33:05 abrighto Exp $"
 *
 * error.C - error reporting routines
 * 
 * who             when      what
 * --------------  --------  ----------------------------------------
 * Allan Brighton  05/10/95  Created
 */
static const char* const rcsId="@(#) $Id: error.C,v 1.2 1998/02/18 11:33:05 abrighto Exp $";


#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <iostream.h>
#include <strstream.h>
#include <errno.h>
#include <stdio.h>
#include "error.h"

// static variable holding text of last error messages
static char errmsg_[5*1024];

// error code (see defines in <sys/errno.h> for values)
static int errno_ = 0;

// optional error handler, to be called with error messages
static void (*errhandler_)(const char*) = NULL;

// optional message handler, to be called with log messages
static void (*msghandler_)(const char*) = NULL;


/*
 * global error reporting routine
 */
int error(const char* msg1, const char* msg2, int code)
{
    char buf[sizeof(errmsg_)];
    ostrstream os(buf, sizeof(buf));
    os << msg1 << msg2 << ends;
    
    if (errhandler_)
	(*errhandler_)(buf);
    else
	print_error(buf);

#ifdef XXXDEBUG
    cerr << "debug: " << buf << endl;
#endif

    errno_ = code;
    strncpy(errmsg_, buf, sizeof(errmsg_)-1);
    return ERROR;
}


/*
 * report the error, including system error code
 */
int sys_error(const char* msg1, const char* msg2)
{
    extern int sys_nerr;
    extern char *sys_errlist[];
    extern int errno;
    char* s = strerror(errno);

    // XXX configure: sys_nerr may not be defined ?
    if (s == NULL || errno < 0 || errno >= sys_nerr)
	return error(msg1, msg2);

    char buf[sizeof(errmsg_)];
    ostrstream os(buf, sizeof(buf));
    os << msg1 << msg2 << ": " << (s ? s : sys_errlist[errno]) << ends;

    if (errhandler_)
	(*errhandler_)(buf);
    else
	print_error(buf);

#ifdef XXXDEBUG
    cerr << "debug: " << buf << endl;
#endif

    errno_ = errno;
    strcpy(errmsg_, buf);
    return ERROR;
}


/*
 * This routine has an interface like printf and reports an error in the
 * same way as error() above.
 */
int fmt_error(const char* fmt, ...)
{
    char buf[1024];
    va_list ap;
    va_start(ap, fmt);
    vsprintf(buf, fmt, ap);
    va_end(ap);
    
    return error(buf);
}


/*
 * This routine has an interface like printf and reports a system error
 * in the same way as sys_error() above.
 */
int fmt_sys_error(const char* fmt, ...)
{
    char buf[1024];
    va_list ap;
    va_start(ap, fmt);
    vsprintf(buf, fmt, ap);
    va_end(ap);
    
    return sys_error(buf);
}


/* 
 * return the text of the previous error message
 */
char* last_error() 
{
    return errmsg_;
}


/* 
 * return the error code for the previous error
 */
int last_error_code() 
{
    return errno_;
}


/* 
 * reset the last_error buf to empty 
 */
void clear_error()
{
    errmsg_[0] = '\0';
}


/*
 * set a routine to be called with the text of error messages
 * when they occur. The argument is a pointer to an error 
 * handler:
 *
 *       void errhandler(const char* msg);
 *
 * The return value is a pointer to the previous error handler, or NULL,
 * if none was defined.
 */
void (*set_error_handler(void (*errhandler)(const char*)))(const char*)
{
    void (*old_handler)(const char*) = errhandler_;
    errhandler_ = errhandler;
    return old_handler;
}


/*  
 * print the given message on stderr (may be used as an error
 * handler)
 */
void print_error(const char* msg)
{
    fprintf(stderr, "%s\n", msg);
    fflush(stderr);
}


/* ---- The routines below are for log messages that are not errors --- */



/*
 * This routine logs a message and has an interface like printf.
 */
void log_message(const char* fmt, ...)
{
    char buf[1024];
    va_list ap;
    va_start(ap, fmt);
    vsprintf(buf, fmt, ap);
    va_end(ap);
    
    if (msghandler_)
	(*msghandler_)(buf);
    else 
	print_log_message(buf);
}

/*
 * set a routine to be called with the text of log messages
 * when they are made. The argument is a pointer to a log message 
 * handler:
 *
 *       void msghandler(const char* msg);
 *
 * The return value is a pointer to the previous message handler, or NULL,
 * if none was defined.
 */
void (*set_log_message_handler(void (*msghandler)(const char*)))(const char*)
{
    void (*old_handler)(const char*) = msghandler_;
    msghandler_ = msghandler;
    return old_handler;
}


/*  
 * print the given message on stdout (may be used as a message
 * handler)
 */
void print_log_message(const char* msg)
{
    printf("%s\n", msg);
    fflush(stdout);
}

