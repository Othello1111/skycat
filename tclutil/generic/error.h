#ifndef _error_h_
#define _error_h_
/*
 * E.S.O. - VLT project 
 * "@(#) $Id: error.h,v 1.2 2005/02/02 01:43:01 brighton Exp $" 
 *
 * error.h - declarations for global error reporting
 * 
 * who             when      what
 * --------------  --------  ----------------------------------------
 * Allan Brighton  05/10/95  Created
 */

/* error status */
#undef OK
#undef ERROR
enum {OK, ERROR};

/* report the error */
int error(const char* msg1, const char* msg2="", int code = 0);

/* report the error with the system error code, like perror */
int sys_error(const char* msg1, const char* msg2="");

/* report error with a printf style interface */
int fmt_error(const char* fmt, ...);

/* report a system error with a printf style interface */
int fmt_sys_error(const char* fmt, ...);

/* return the text of the previous error message */
char* last_error();

/* return the error code for the previous error */
int last_error_code();

/* reset the last_error buf to empty */
void clear_error();

/* set a routine to be called when errors occur and return ptr to previous handler */
void (*set_error_handler(void (*error_handler)(const char*)))(const char*);

/* simple error handler: print the message */
void print_error(const char* msg);

/* print a message (like printf) */
void log_message(const char* fmt, ...);

/* set a routine to be called for log messages and return ptr to previous handler */
void (*set_log_message_handler(void (*message_handler)(const char*)))(const char*);

/* simple message handler: print the message */
void print_log_message(const char* msg);



#endif /* _error_h_ */

