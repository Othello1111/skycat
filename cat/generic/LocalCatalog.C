/*
 * E.S.O. - VLT project/ESO Archive
 * $Id: LocalCatalog.C,v 1.1.1.1 2009/03/31 14:11:52 cguirao Exp $
 *
 * LocalCatalog.C - method definitions for class LocalCatalog
 * 
 * See the man page for a complete description.
 * 
 * who             when       what
 * --------------  --------   ----------------------------------------
 * Allan Brighton  11 Jun 96  Created
 */
static const char* const rcsId="@(#) $Id: LocalCatalog.C,v 1.1.1.1 2009/03/31 14:11:52 cguirao Exp $";


#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <cstring>
#include "error.h"
#include "Mem.h"
#include "LocalCatalog.h"


/* 
 * static method to check the validity of a tab table file.
 * Returns 0 if OK.
 */
int LocalCatalog::check_table(const char* file)
{
    TabTable t;
    return TabTable::head(file, t);
}


/*
 * constructor - used internally only, public interface uses "open(name)" 
 * "e" is the catalog config entry object for this catalog.
 * (see CatalogInfo class)
 *
 * In this case, the catalog config entry may have been created automatically.
 * The name of the file containing the local catalog (tab table) is stored in
 * "e->url()", which is normally used to store the URL for HTTP access, but is
 * used here for the filename.
 *
 */
LocalCatalog::LocalCatalog(CatalogInfoEntry* e) 
    : AstroCatalog(e),
      filename_(strdup(e->url()))
{
    status_ = getInfo();
}


/*
 * destructor
 */
LocalCatalog::~LocalCatalog() 
{
    if (filename_)
	free(filename_);
}


/*
 * Run a query on the local catalog and return the number of objects found.
 *
 * Args:
 *     q -        (in)   object describing the query
 *
 *     filename - (in)   filename to hold results, or null
 *
 *     result -   (out)  reference to object to manage the results. 
 *
 * The return value is the number of rows found, or 0 if none were found.
 * A return value of -1 indicates an error.
 *
 * (Redefined from parent class to work with local catalogs)
 */
int LocalCatalog::query(const AstroQuery& q, const char* filename, QueryResult& result)
{
    if (checkInfo() != 0)
	return -1;

    // note the catalog config entry in the results
    result.entry(entry_);

    if (result.query(q, info_, filename, more_) != 0)
	return -1;

    return result.numRows();
}


/*
 * If we don't have the info for this catalog, get it and 
 * return the status. Here we also check if the file has been modified,
 * (by an insert or remove operation) and reload it if needed.
 */
int LocalCatalog::checkInfo()
{
    if (info_.numCols() > 0) {
	struct stat buf;
	if (stat(filename_, &buf) != 0)
	    return sys_error("can't access file: ", filename_);
	if (buf.st_mtime == timestamp_)
	    return 0;
    }
    return getInfo();
}


/*
 * Read the local catalog to get the column names and also read in the
 * data to make later searches faster.  The return value is 0 for
 * success. The info_ member holds the column info and the local catalog
 * data for searching. It must be updated if the data changes.
 */
int LocalCatalog::getInfo()
{
    // note update time of file, so we know if it has been modified...
    struct stat buf;
    if (stat(filename_, &buf) != 0)
	return sys_error("can't access file: ", filename_);
    timestamp_ = buf.st_mtime;

    // mmap the file and put it in a TabTable
    Mem m(filename_);
    if (m.status() != 0)
	return 1;
    
    // make a null terminated copy, which will be managed by info_
    size_t size = m.size() + 1;
    char* data = (char*)malloc(size);
    if (!data)
	return fmt_error("can't allocate %d bytes for %s", size, filename_);
    strncpy(data, (char*)m.ptr(), size-1);
    data[size-1] = '\0';

    if (info_.init(data, 0, 1) != 0)
	return 1;

    // copy the comments from table to entry
    int n = info_.numComments();
    if ( n > 0 ) {
        char* c = NULL;
        int have = 1024;
        int l = 0;
        int used = 0;

        char* com = (char*)malloc(have);
        com[0] = '\0';
        char* p = com;

        for ( int i = 0; i < n; i++ ) {
            info_.getComment( i, c );
            l = strlen( c );
            if (( used + l ) >= have) { 
                have += 1024;
                com = (char*)realloc(com, have);
                p = com + used;
            }
            strcpy( p, c );
            used += l+1;
            p += l;
            if (i < (n-1)) {
                *p++ = '\n'; //  new line not NULL.
            }
        }
        entry_->comments(com);
        free(com);
    }

    // this will extract any catalog config info from the file's header
    info_.entry(entry_, data);
    return 0;
}

