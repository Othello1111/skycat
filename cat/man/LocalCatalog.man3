# E.S.O. - VLT project/ESO Archive
# $Id: LocalCatalog.man3,v 1.1.1.1 2006/01/12 16:36:07 abrighto Exp $
#
# This file is processed by the ESO/VLT docDoManPages command to
# produce a man page in nroff, TeX and MIF formats. 
# See docDoManPages(1) for a description of the input format.
# 
# who             when       what
# --------------  ---------  ----------------------------------------
# Allan Brighton  27 Jun 96  Created
#

NAME
LocalCatalog - Class for accessing local catalogs stored as tab tables

SYNOPSIS
#include "LocalCatalog.h"

class LocalCatalog : public AstroCatalog {
...
public:
    LocalCatalog(const CatalogInfoEntry*);

    virtual int query(const AstroQuery& q, const char* filename, QueryResult& result);

    static int check_table(const char* file);

    virtual int sort(const char* column, int numeric = 1);

    virtual int uniq();
};

DESCRIPTION
This class can be used to search a local catalog stored as a file in
tab table format. The format of the table is something like:

 TableName

 # comment line 

 # catalog config information
 keyword: value ...


 A	B	C
 -	-	-
 0	1	3
 3	2	4
 ...

where the table name and catalog config info are optional. This is
the same format used by the "Starbase" package or utilities (see:
http://cfa-www.harvard.edu/~john/ for more about Starbase).

UNITS
Unless otherwise stated, the units for all radius values are in arcmin
or image pixels, depending on the catalog config entry settings. The
default is arcmin.  Floating point values for RA and DEC are always in
degrees. The default equinox is J2000, unless set otherwise in the
catalog config entry or in the table header.

PUBLIC METHODS

LocalCatalog(const CatalogInfoEntry*)
	Constructor - create local catalog class instance.  
	Note: public interface normally uses AstroCatalog::open() with
	the name of the file containing the local catalog.  The
	argument represents the entry in the catalog config file for
	this catalog (made automatially, if not already present).

query(const AstroQuery& q, const char* filename, QueryResult& result)
	Run a query on the catalog and return the number of objects
	found (redefined here to work with local catalogs, refer to
	the parent class AstroCatalog for details.).

check_table(const char* file)
	Check the validity of a tab table file and return 0 if it is ok.

SEE ALSO

AstroCatalog
----------------------------------------------------------------------
