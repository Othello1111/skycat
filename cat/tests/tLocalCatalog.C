/*
 * E.S.O. - VLT project / ESO Archive
 * $Id: tLocalCatalog.C,v 1.1.1.1 2006/01/12 16:36:12 abrighto Exp $
 *
 * tLocalCatalog.C - test cases for class LocalCatalog
 * 
 * who             when       what
 * --------------  --------   ----------------------------------------
 * Allan Brighton  11 Jun 96  Created
 */

using namespace std;
#include <stdio.h>
#include <iostream>
#include <stdlib.h>
#include "error.h"
#include "LocalCatalog.h"

main() 
{
    // gcc doesn't need this, but SunCC does...
    ios::sync_with_stdio();

    // errors will be printed on stderr automatically
    set_error_handler(print_error);

   // open the GSC catalog
    AstroCatalog* cat = AstroCatalog::open("test.table");
    if (!cat) {
	printf("LocalCatalog::open returned NULL\n");
	exit(1);
    }

    // query the GSC for a list of objects at the given pos and radius
    WorldCoords pos(3, 19, 48, 41, 30, 39);
    AstroQuery q;
    q.pos(pos);
    q.radius(10);
    q.maxRows(10);

    cout << "Query GSC for objects at pos " 
	 << q.pos() << ", in radius " 
	 << q.radius1() << ".." << q.radius2() 
	 << ":" 
	 << endl;

    QueryResult result;
    int num_results = cat->query(q, "./tLocalCatalog.out", result);
    if (num_results < 0) {
	cout << "Query returned an error\n";
	exit(1);
    }

    cout << "Query returns: " 
	 << num_results 
	 << " objects, out of " 
	 << q.maxRows() 
	 << " requested" 
	 << endl;

    cout << "More objects ?: " 
	 << (cat->more() ? "YES" : "NO") 
	 << endl;

    cout << "---Result List---" 
	 << endl;

    int ncols = cat->numCols();
    int i, j;
    cout << "columns:\n";
    for (i = 0; i < ncols; i++) 
	cout << cat->colName(i) << "\t";
    cout << "\n\nresults:\n";

    char* s;
    for (i = 0; i < num_results; i++) {
	// print the Id
	if (result.get(i, 0, s) == 0)
	    cout << s << "\t";
	else
	    cout << "ERROR\t";
	    
	// print the position in H:M:S+D:M:S
	if (result.getPos(i, pos) == 0)
	    cout << pos.ra() << "\t" << pos.dec() << "\t";
	else
	    cout << "ERROR\t";

	// print the other columns
	for (j = 3; j < ncols; j++) {
	    if (result.get(i, j, s) == 0)
		cout << s << "\t";
	    else
		cout << "ERROR\t";
	}
	cout << endl;
    }
    cout << "-----------------" << endl;

    cout << "test searching by Id in a local catalog\n";
    if (cat->getObject("GSC0286902884",	0, NULL, result) != 0) {
	cout << "ERROR\n";
    }
    else {
	cout << "OK\n" << result << endl;;
    }

    
    return 0;
}
