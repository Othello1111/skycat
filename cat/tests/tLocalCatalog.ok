Query GSC for objects at pos 03:19:48.000 +41:30:39.00, in radius 0..10:
Query returns: 10 objects, out of 10 requested
More objects ?: YES
---Result List---
columns:
gsc-id	ra	dec	pos-e	mag	mag-e	mu	d'	pa	

results:
GSC0285600050	03:19:28.140	+41:26:29.00	0.2	13.49	0.40	F	5.63	222;	
GSC0285600098	03:19:32.990	+41:34:15.42	0.4	13.40	0.40	F	4.58	322;	
GSC0285600288	03:19:21.100	+41:31:19.74	0.2	13.13	0.40	F	5.12	277;	
GSC0285600314	03:19:44.440	+41:30:58.21	0.2	13.98	0.40	F	0.77	293;	
GSC0285600334	03:19:48.200	+41:32:50.14	0.2	11.77	0.40	F	2.16	360;	
GSC0285600342	03:19:57.260	+41:31:34.72	0.2	12.31	0.40	F	1.92	62;	
GSC0285600418	03:19:24.110	+41:34:44.58	0.2	11.69	0.40	F	6.08	312;	
GSC0285600518	03:19:47.840	+41:25:46.45	0.2	12.74	0.40	F	4.90	181;	
GSC0285600526	03:19:26.740	+41:32:26.23	0.2	11.25	0.40	F	4.39	294;	
GSC0285600548	03:19:51.070	+41:26:09.89	0.4	12.09	0.40	F	4.54	173;	
-----------------
test searching by Id in a local catalog
OK
QueryResult

# Config entry for original catalog server:
serv_type: local
long_name: test.table
short_name: test.table
url: test.table
# End config entry

gsc-id	ra	dec	pos-e	mag	mag-e	mu	d'	pa
------	--	---	-----	---	-----	--	--	--
GSC0286902884	03:20:0.87	+41:33:14.00	0.2	13.07	0.40	F	3.49	43;

