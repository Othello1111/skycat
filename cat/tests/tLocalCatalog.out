QueryResult

# Config entry for original catalog server:
serv_type: local
long_name: test.table
short_name: test.table
url: test.table
# End config entry

gsc-id	ra	dec	pos-e	mag	mag-e	mu	d'	pa
------	--	---	-----	---	-----	--	--	--
GSC0285600050	03:19:28.14	+41:26:29.00	0.2	13.49	0.40	F	5.63	222;
GSC0285600098	03:19:32.99	+41:34:15.42	0.4	13.40	0.40	F	4.58	322;
GSC0285600288	03:19:21.10	+41:31:19.74	0.2	13.13	0.40	F	5.12	277;
GSC0285600314	03:19:44.44	+41:30:58.21	0.2	13.98	0.40	F	0.77	293;
GSC0285600334	03:19:48.20	+41:32:50.14	0.2	11.77	0.40	F	2.16	360;
GSC0285600342	03:19:57.26	+41:31:34.72	0.2	12.31	0.40	F	1.92	62;
GSC0285600418	03:19:24.11	+41:34:44.58	0.2	11.69	0.40	F	6.08	312;
GSC0285600518	03:19:47.84	+41:25:46.45	0.2	12.74	0.40	F	4.90	181;
GSC0285600526	03:19:26.74	+41:32:26.23	0.2	11.25	0.40	F	4.39	294;
GSC0285600548	03:19:51.07	+41:26:9.89	0.4	12.09	0.40	F	4.54	173;
